File: src/main/java/net/snipsniper/utils/InfoButton.java
package net.snipsniper.utils;

import net.snipsniper.ImageManager;

import javax.swing.*;
import java.awt.*;
import java.awt.event.*;

public class InfoButton extends JButton {
    private final String info;
    private JFrame window;

    public InfoButton(String info) {
        if(info == null) {
            this.info = "No text provided";
        } else {
            this.info = info;
        }
        setText("?");

        addMouseListener(new MouseAdapter() {
            @Override
            public void mouseEntered(MouseEvent mouseEvent) {
                if(!isEnabled())
                    return;
                super.mouseEntered(mouseEvent);
                new java.util.Timer().schedule(new java.util.TimerTask() {
                    @Override
                    public void run() {
                        if(isShowing()) {
                            Rectangle rect = new Rectangle(getLocationOnScreen().x, getLocationOnScreen().y, getBounds().width, getBounds().height);
                            if (rect.contains(MouseInfo.getPointerInfo().getLocation())) {
                                openWindow();
                            }
                        }
                    }
                }, 1000);
            }

            @Override
            public void mouseExited(MouseEvent mouseEvent) {
                super.mouseExited(mouseEvent);
                closeWindow();
            }
        });
    }

    @Override
    public void paint(Graphics g) {
        if(!isEnabled())
            return;
        g.setColor(new Color(0, 0, 0, 0));
        g.drawRect(0, 0, getWidth(), getHeight());
        int iconSize = 16;
        g.drawImage(ImageManager.getImage("icons/questionmark.png").getScaledInstance(iconSize, iconSize, Image.SCALE_SMOOTH), getWidth() /2 - iconSize / 2, getHeight() / 2 - iconSize / 2, iconSize, iconSize, this);
    }

    public void closeWindow() {
        if(window != null) {
            window.dispose();
            window = null;
        }
    }

    public void openWindow() {
        if(window != null) {
            window.requestFocus();
            return;
        }
        window = new JFrame();
        window.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
        window.setLocation(getLocationOnScreen().x + getWidth(), getLocationOnScreen().y);
        window.getContentPane().setBackground(Color.WHITE);
        window.setUndecorated(true);
        JLabel content = new JLabel("<html><p style=\"width:256px;\">" + info + "</p></html>");
        content.setVerticalAlignment(JLabel.TOP);
        window.add(content);
        window.setMinimumSize(new Dimension(256, 128));
        window.getRootPane().setBorder(BorderFactory.createMatteBorder(1, 1, 1, 1, Color.BLACK));
        window.addKeyListener(new KeyAdapter() {
            @Override
            public void keyReleased(KeyEvent keyEvent) {
                if(keyEvent.getKeyCode() == KeyEvent.VK_ESCAPE)
                    window.dispose();
            }
        });
        window.setVisible(true);
        window.pack();
    }
}

File: src/main/kotlin/net/snipsniper/utils/InfoButton.kt
package net.snipsniper.utils

import java.awt.*
import java.awt.event.*
import java.util.*
import javax.swing.BorderFactory
import javax.swing.JButton
import javax.swing.JFrame
import javax.swing.JLabel

class InfoButton(): JButton() {
    private lateinit var info: String
    private var window: JFrame? = null

    constructor(info: String?) : this() {
        this.info = info ?: "No text provided"
        text = "?"
        addMouseListener(object: MouseAdapter() {
            override fun mouseEntered(e: MouseEvent?) {
                if(!isEnabled) return
                super.mouseEntered(e)
                Timer().schedule(object: TimerTask() {
                    override fun run() {
                        if(isShowing) {
                            Rectangle(locationOnScreen.x, locationOnScreen.y, bounds.width, bounds.height).also {
                                if(it.contains(MouseInfo.getPointerInfo().location))
                                    openWindow()
                            }
                        }
                    }
                }, 1000)
            }

            override fun mouseExited(e: MouseEvent?) {
                super.mouseExited(e)
                closeWindow()
            }
        })
    }

    override fun paint(g: Graphics) {
        if(!isEnabled) return
        g.color = Color(0, 0, 0, 0)
        g.drawRect(0, 0, width, height)
        val iconSize = 16
        fun fk(n: Int): Int = n / 2 - iconSize / 2
        g.drawImage("icons/questionmark.png".getImage(iconSize, iconSize), fk(width), fk(height), iconSize, iconSize, this)
    }

    fun closeWindow() {
        window?.dispose()
        window = null
    }

    fun openWindow() {
        if(window != null) {
            window?.requestFocus()
            return
        }

        window = JFrame().also {
            it.defaultCloseOperation = JFrame.DISPOSE_ON_CLOSE
            it.setLocation(locationOnScreen.x + width, locationOnScreen.y)
            it.contentPane.background = Color.WHITE
            it.isUndecorated = true
            JLabel("<html><p style=\"width:256px;\">$info</p></html>").also { label ->
                label.verticalAlignment = JLabel.TOP
                it.add(label)
            }
            it.minimumSize = Dimension(256, 128)
            it.rootPane.border = BorderFactory.createMatteBorder(1, 1, 1, 1, Color.BLACK)
            it.addKeyListener(object: KeyAdapter() {
                override fun keyReleased(e: KeyEvent) {
                    if(e.keyCode == KeyEvent.VK_ESCAPE)
                        it.dispose()
                }
            })
            it.isVisible = true
            it.pack()
        }
    }

}